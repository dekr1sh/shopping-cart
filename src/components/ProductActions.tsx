import { Button } from "@/components/ui/button"
import { useCart } from "@/context/useCart"
import type { Product } from "@/data/products"
import { useState } from "react"
import { toast } from "sonner"

export default function ProductActions({ product }: { product: Product }) {
  const { addToCart } = useCart()
  const [selectedSize, setSelectedSize] = useState("")
  const [quantity, setQuantity] = useState(1)
  const [error, setError] = useState("")

  const sizes = ["7", "7.5", "8", "8.5", "9", "9.5", "10", "10.5", "11", "11.5", "12"]

  const handleAddToCart = () => {
    if (!selectedSize) {
      setError("Please select a size")
      return
    }

    setError("")
    addToCart(product, selectedSize, quantity)

    toast.success("Added to cart", {
      description: `${product.name} (Size: ${selectedSize}) has been added to your cart.`,
      duration: 3000,
    })
  }

  return (
    <>
      <div className="mb-6">
        <h3 className="font-medium mb-2">Select Size</h3>
        <div className="grid grid-cols-4 gap-2">
          {sizes.map((size) => (
            <button
              key={size}
              className={`py-2 px-4 rounded border cursor-pointer ${selectedSize === size
                  ? "bg-primary text-primary-foreground border-primary"
                  : "bg-background text-foreground border-border hover:border-primary"
                }`}
              onClick={() => setSelectedSize(size)}
            >
              {size}
            </button>
          ))}
        </div>
        {error && <p className="text-sm text-destructive mt-2">{error}</p>}
      </div>

      <div className="mb-6">
        <h3 className="font-medium mb-2">Quantity</h3>
        <div className="flex items-center border border-border rounded w-32">
          <button
            className="px-3 py-1 hover:bg-muted cursor-pointer"
            onClick={() => setQuantity((prev) => Math.max(1, prev - 1))}
            disabled={quantity <= 1}
          >
            -
          </button>
          <input
            type="number"
            min="1"
            value={quantity}
            onChange={(e) => setQuantity(+e.target.value || 1)}
            className="w-full text-center py-1 border-0 focus:ring-0 bg-transparent"
          />
          <button className="px-3 py-1 hover:bg-muted cursor-pointer" onClick={() => setQuantity((prev) => prev + 1)}>
            +
          </button>
        </div>
      </div>

      <Button className="w-full cursor-pointer" size="lg" onClick={handleAddToCart}>
        Add to Cart
      </Button>
    </>
  )
}
