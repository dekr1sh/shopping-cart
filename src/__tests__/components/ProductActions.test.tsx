import { screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { vi, describe, it, expect } from "vitest"
import { render } from "@/__tests__/test-utils"
import ProductActions from "@/components/ProductActions"

const mockAddToCart = vi.fn()

vi.mock("@/context/useCart", () => ({ useCart: vi.fn(() => ({addToCart: mockAddToCart}))}));

const mockProduct = {
  id: 1,
  name: "Mock Sneaker",
  brand: "Mock Brand",
  price: 1999,
  description: "A cool pair of sneakers.",
  image: "/mock.jpg",
}

describe("ProductActions", () => {
  it("renders correctly when no size is selected", () => {
    const { container } = render(<ProductActions product={mockProduct} />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly after selecting a size", async () => {
    const user = userEvent.setup();

    const { container } = render(<ProductActions product={mockProduct} />)
    await user.click(screen.getByRole("button", { name: "9" }))

    expect(container).toMatchSnapshot()
  })

  it("does not add to cart if size is not selected", async () => {
    const user = userEvent.setup();

    render(<ProductActions product={mockProduct} />)
    await user.click(screen.getByRole("button", { name: /add to cart/i }))

    expect(mockAddToCart).not.toHaveBeenCalled()
  })

  it("adds to cart when size is selected", async () => {
    const user = userEvent.setup();

    render(<ProductActions product={mockProduct} />)

    await user.click(screen.getByRole("button", { name: "10" }))
    await user.click(screen.getByRole("button", { name: /add to cart/i }))

    expect(mockAddToCart).toHaveBeenCalledWith(mockProduct, "10", 1)
  })

  it("adds to cart with correct quantity", async () => {
    const user = userEvent.setup()
    render(<ProductActions product={mockProduct} />)
  
    await user.click(screen.getByRole("button", { name: "9" }))
    await user.click(screen.getByRole("button", { name: "+" })) 
  
    await user.click(screen.getByRole("button", { name: /add to cart/i }))
    expect(mockAddToCart).toHaveBeenCalledWith(mockProduct, "9", 2)
  })
})
