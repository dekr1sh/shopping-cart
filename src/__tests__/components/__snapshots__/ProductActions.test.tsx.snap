// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductActions > renders correctly after selecting a size 1`] = `
<div>
  <div
    class="mb-6"
  >
    <h3
      class="font-medium mb-2"
    >
      Select Size
    </h3>
    <div
      class="grid grid-cols-4 gap-2"
    >
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        7
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        7.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        8
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        8.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-primary text-primary-foreground border-primary"
      >
        9
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        9.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        10
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        10.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        11
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        11.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        12
      </button>
    </div>
  </div>
  <div
    class="mb-6"
  >
    <h3
      class="font-medium mb-2"
    >
      Quantity
    </h3>
    <div
      class="flex items-center border border-border rounded w-32"
    >
      <button
        class="px-3 py-1 hover:bg-muted cursor-pointer"
        disabled=""
      >
        -
      </button>
      <input
        class="w-full text-center py-1 border-0 focus:ring-0 bg-transparent"
        min="1"
        type="number"
        value="1"
      />
      <button
        class="px-3 py-1 hover:bg-muted cursor-pointer"
      >
        +
      </button>
    </div>
  </div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-10 rounded-md px-6 has-[>svg]:px-4 w-full cursor-pointer"
    data-slot="button"
  >
    Add to Cart
  </button>
     
</div>
`;

exports[`ProductActions > renders correctly when no size is selected 1`] = `
<div>
  <div
    class="mb-6"
  >
    <h3
      class="font-medium mb-2"
    >
      Select Size
    </h3>
    <div
      class="grid grid-cols-4 gap-2"
    >
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        7
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        7.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        8
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        8.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        9
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        9.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        10
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        10.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        11
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        11.5
      </button>
      <button
        class="py-2 px-4 rounded border cursor-pointer bg-background text-foreground border-border hover:border-primary"
      >
        12
      </button>
    </div>
  </div>
  <div
    class="mb-6"
  >
    <h3
      class="font-medium mb-2"
    >
      Quantity
    </h3>
    <div
      class="flex items-center border border-border rounded w-32"
    >
      <button
        class="px-3 py-1 hover:bg-muted cursor-pointer"
        disabled=""
      >
        -
      </button>
      <input
        class="w-full text-center py-1 border-0 focus:ring-0 bg-transparent"
        min="1"
        type="number"
        value="1"
      />
      <button
        class="px-3 py-1 hover:bg-muted cursor-pointer"
      >
        +
      </button>
    </div>
  </div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-10 rounded-md px-6 has-[>svg]:px-4 w-full cursor-pointer"
    data-slot="button"
  >
    Add to Cart
  </button>
     
</div>
`;
